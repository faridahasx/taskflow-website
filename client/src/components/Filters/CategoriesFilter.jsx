// External imports
import PropTypes from "prop-types";
// MUI components
import InputLabel from "@mui/material/InputLabel";
import MenuItem from "@mui/material/MenuItem";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import Checkbox from "@mui/material/Checkbox";
import ListItemText from "@mui/material/ListItemText";

// Menu props
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: 48 * 4.5 + 8,
      width: 250,
    },
  },
};

const CategoriesFilter = (props) => {
  // Destructuring props
  const { categories, selectedCategories, setSelectedCategories } = props;

  // Handle change in category selection
  const handleChange = (e) => {
    e.preventDefault();
    let {
      target: { value },
    } = e;
    value = typeof value === "string" ? value.split(",") : value;
    // Update selected categories based on the value
    const updatedCategories =
      value[value.length - 1] === "All"
        ? ["All"]
        : value[0] === "All"
        ? [value[1]]
        : value;
    // Set the updated categories in the state
    setSelectedCategories(updatedCategories);
  };

  return (
    <div className="sf-component">
      <InputLabel id="category-filter-label">Select categories:</InputLabel>
      <FormControl sx={{ width: "100%" }}>
        <Select
          labelId="category-filter-label"
          displayEmpty
          id="filter-category"
          multiple
          value={selectedCategories}
          onChange={handleChange}
          renderValue={() => selectedCategories.join(", ")}
          MenuProps={MenuProps}
        >
          {categories.map((c) => (
            <MenuItem key={c.title} value={c.title} className="portal">
              <Checkbox
                checked={selectedCategories.indexOf(c.title) > -1}
                className="portal"
              />
              <ListItemText primary={c.title} className="portal" />
            </MenuItem>
          ))}
        </Select>
      </FormControl>
    </div>
  );
};

CategoriesFilter.propTypes = {
  categories: PropTypes.array.isRequired,
  selectedCategories: PropTypes.array.isRequired,
  setSelectedCategories: PropTypes.func.isRequired,
};

export default CategoriesFilter;
