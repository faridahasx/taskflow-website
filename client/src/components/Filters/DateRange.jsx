import PropTypes from "prop-types";
// MUI components
import InputLabel from "@mui/material/InputLabel";
// Components
import DatePicker from "../DatePicker";

const DateRange = (props) => {
  const { start, end, setStart, setEnd, label } = props;
  return (
    <div className="sf-component">
      <InputLabel id="category-filter-label">{label}</InputLabel>
      <DatePicker
        name="Start"
        label="Start"
        date={start}
        setDate={setStart}
        sx={{ width: "50%" }}
      />
      <DatePicker
        name="End"
        label="End"
        date={end}
        setDate={setEnd}
        sx={{ width: "50%", paddingLeft: "5px" }}
      />
    </div>
  );
};

DateRange.propTypes = {
  start: PropTypes.any,
  end: PropTypes.any,
  setStart: PropTypes.func.isRequired,
  setEnd: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
};

export default DateRange;
