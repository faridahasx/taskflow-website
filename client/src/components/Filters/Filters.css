#filters-form {
    width: calc(100vw - 50px);
    max-width: 550px;
    min-width: calc(var(--min-width) - 50px);
    max-height: calc(var(--vh) - 100px);
    overflow: hidden;
}

#filters-form-fields {
    padding: 10px;
    max-height: calc(var(--vh) - 200px);
    overflow: auto;
}


#category-filter-label {
    padding: 5px 10px;
    color: rgb(63, 63, 63);
}

#target-label {
    padding: 10px;
    color: rgb(63, 63, 63);
}

#date-filter {
    width: 100%;
}


#completed-filters label {
    width: 100%;
    padding: 0 10px;
}

#completed-filters label,
#completed-filters input {
    color: rgb(63, 63, 63);
}

/* Bottom */
#filters-bottom {
    padding: 10px;
}

#filters-bottom button:first-child {
    font-weight: 500;

}