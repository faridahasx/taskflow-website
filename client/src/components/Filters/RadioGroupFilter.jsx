import PropTypes from "prop-types";
// MUI components
import Radio from "@mui/material/Radio";
import RadioGroup from "@mui/material/RadioGroup";
import FormControlLabel from "@mui/material/FormControlLabel";

const RadioGroupFilter = ({
  value,
  onChange,
  name,
  option1Label,
  option2Label,
}) => {
  return (
    <div className="sf-component">
      <RadioGroup value={value} onChange={onChange} name={name}>
        <FormControlLabel
          value={true}
          control={<Radio />}
          label={option1Label}
        />
        <FormControlLabel
          value={false}
          control={<Radio />}
          label={option2Label}
        />
        <FormControlLabel value="" control={<Radio />} label="All" />
      </RadioGroup>
    </div>
  );
};

RadioGroupFilter.propTypes = {
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  option1Label: PropTypes.string.isRequired,
  option2Label: PropTypes.string.isRequired,
};

export default RadioGroupFilter;
