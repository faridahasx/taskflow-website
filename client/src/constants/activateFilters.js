const activateFilters = [
  { buttonTxt: "All", value: "all" },
  { buttonTxt: "Current", value: "current" },
  { buttonTxt: "Overdue", value: "overdue" },
  { buttonTxt: "Upcoming", value: "upcoming" },
  { buttonTxt: "Completed", value: "completed" },
];

export default activateFilters;
