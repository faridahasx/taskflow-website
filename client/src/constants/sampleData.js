import { getDateWithNDaysOfDifference } from "../utils/dateUtils";

const categoriesSample = [
  {
    title: "All",
    tasks: 5,
    _id: "All",
  },
  {
    _id: "Tasks",
    tasks: 5,
    title: "Tasks",
  },
];

let tasksSample = [
  {
    _id: "1",
    title: "Task 1",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-1),
    finishDate: getDateWithNDaysOfDifference(1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "2",
    title: "Task 2",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-2),
    finishDate: getDateWithNDaysOfDifference(-1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "3",
    title: "Task 3",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(2),
    finishDate: getDateWithNDaysOfDifference(3),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },

  {
    _id: "4",
    title: "Task 4",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(2),
    finishDate: getDateWithNDaysOfDifference(4),
    completedAt: getDateWithNDaysOfDifference(3),
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
  {
    _id: "5",
    title: "Login to continue",
    category: "Tasks",
    description: "",
    startDate: getDateWithNDaysOfDifference(-1),
    finishDate: getDateWithNDaysOfDifference(1),
    completedAt: null,
    createdAt: getDateWithNDaysOfDifference(-4),
    updatedAt: getDateWithNDaysOfDifference(-4),
    __v: 0,
  },
];

export { tasksSample, categoriesSample };
