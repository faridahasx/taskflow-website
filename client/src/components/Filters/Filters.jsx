// External imports
import PropTypes from "prop-types";
// Custom hooks
import useBodyOverflowHidden from "../../hooks/useBodyOverflowHidden";
// Components
import Modal from "../Modal/Modal";
import PortalComponent from "../PortalComponent";
import SubmitButton from "../TextButtons/SubmitButton";
import CancelButton from "../TextButtons/CancelButton";
import SortAndFiltersWrapper from "../SortAndFiltersWrapper/SortAndFiltersWrapper";
import CategoriesFilter from "./CategoriesFilter";
import DateRange from "./DateRange";
// Styles
import "./Filters.css";

const Filters = (props) => {
  // Destructuring props
  const {
    handleClose,
    handleSubmit,
    categories,
    selectedCategories,
    setSelectedCategories,
    startgte,
    setStartgte,
    startlte,
    setStartlte,
    finishgte,
    setFinishgte,
    finishlte,
    setFinishlte,
    handleClear,
  } = props;

  useBodyOverflowHidden();

  // Render component
  return (
    <PortalComponent>
      <Modal handleClose={handleClose}>
        <SortAndFiltersWrapper handleClose={handleClose} heading="Filters">
          <form id="filters-form" onSubmit={handleSubmit} data-testid="filters">
            <div id="filters-form-fields">
              <CategoriesFilter
                categories={categories}
                selectedCategories={selectedCategories}
                setSelectedCategories={setSelectedCategories}
              />
              <DateRange
                label="Start range"
                start={startgte}
                end={startlte}
                setStart={setStartgte}
                setEnd={setStartlte}
              />
              <DateRange
                label="Finish range"
                start={finishgte}
                end={finishlte}
                setStart={setFinishgte}
                setEnd={setFinishlte}
              />
            </div>
            <div id="filters-bottom" className="flex">
              <CancelButton buttonText="Clear Filters" onClick={handleClear} />
              <SubmitButton
                buttonText="Apply"
                buttonProps={{ form: "filters-form" }}
              />
            </div>
          </form>
        </SortAndFiltersWrapper>
      </Modal>
    </PortalComponent>
  );
};

Filters.propTypes = {
  handleClose: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  setSelectedCategories: PropTypes.func.isRequired,
  setStartgte: PropTypes.func.isRequired,
  setStartlte: PropTypes.func.isRequired,
  setFinishgte: PropTypes.func.isRequired,
  setFinishlte: PropTypes.func.isRequired,
  handleClear: PropTypes.func.isRequired,
  categories: PropTypes.array,
  selectedCategories: PropTypes.array,
  startgte: PropTypes.any,
  startlte: PropTypes.any,
  finishgte: PropTypes.any,
  finishlte: PropTypes.any,
};

export default Filters;
