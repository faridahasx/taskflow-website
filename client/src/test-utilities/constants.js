import { MISSING_INPUT_FIELD } from "constants/alertMessages";

// AUTH
export const passwordLabel = "Password:";
export const emailLabel = "Email:";
export const firstnameLabel = "First Name:";
export const lastnameLabel = "Last Name:";

// Server responses
export const successResponse = { status: 200, data: "Success" };

// Redux state
export const reduxStateOnMissingInputField = [
  {
    type: "ALERT",
    payload: MISSING_INPUT_FIELD,
  },
];
