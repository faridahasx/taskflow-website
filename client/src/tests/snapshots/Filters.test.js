import { render } from "@testing-library/react";
import Filters from "../../components/Filters/Filters";
import { categoriesSample } from "../../constants/sampleData";

const mockFn = jest.fn();

test("Filters snapshot", () => {
  const { asFragment } = render(
    <Filters
      handleClose={mockFn}
      handleSubmit={mockFn}
      setSelectedCategories={mockFn}
      setStartgte={mockFn}
      setStartlte={mockFn}
      setFinishgte={mockFn}
      setFinishlte={mockFn}
      handleRadioGroupChange={mockFn}
      handleClear={mockFn}
      categories={categoriesSample}
    />
  );
  expect(asFragment()).toMatchSnapshot("Filters");
});
