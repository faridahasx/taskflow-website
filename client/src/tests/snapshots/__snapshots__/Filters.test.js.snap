// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters snapshot: Filters 1`] = `<DocumentFragment />`;
